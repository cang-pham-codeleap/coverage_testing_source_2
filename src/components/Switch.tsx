import React, { useState } from 'react';

interface SwitchProps {
  /**
   * Whether the switch is checked
   */
  checked?: boolean;
  
  /**
   * Default checked state (uncontrolled component)
   */
  defaultChecked?: boolean;
  
  /**
   * Disabled state of the switch
   */
  disabled?: boolean;
  
  /**
   * Function called when the switch state changes
   */
  onChange?: (checked: boolean) => void;
  
  /**
   * Size of the switch: 'sm' (small), 'md' (medium), or 'lg' (large)
   */
  size?: 'sm' | 'md' | 'lg';
  
  /**
   * Additional CSS class names
   */
  className?: string;
  
  /**
   * Label for accessibility
   */
  ariaLabel?: string;
}

/**
 * Switch component for toggling between two states
 * 
 * @example
 * ```tsx
 * <Switch checked={isActive} onChange={setIsActive} />
 * <Switch defaultChecked size="lg" />
 * ```
 */
const Switch: React.FC<SwitchProps> = ({
  checked,
  defaultChecked = false,
  disabled = false,
  onChange,
  size = 'md',
  className = '',
  ariaLabel,
}) => {
  // For uncontrolled component usage
  const [internalChecked, setInternalChecked] = useState<boolean>(defaultChecked);
  
  // Determine if component is controlled or uncontrolled
  const isControlled = checked !== undefined;
  const isChecked = isControlled ? checked : internalChecked;
  
  // Size-based styling
  const sizeClasses = {
    sm: {
      wrapper: 'w-8 h-4',
      thumb: 'w-3 h-3',
      thumbTranslate: 'translate-x-4',
    },
    md: {
      wrapper: 'w-10 h-5',
      thumb: 'w-4 h-4',
      thumbTranslate: 'translate-x-5',
    },
    lg: {
      wrapper: 'w-12 h-6',
      thumb: 'w-5 h-5',
      thumbTranslate: 'translate-x-6',
    },
  };
  
  // Handle change
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    // Update internal state for uncontrolled usage
    if (!isControlled) {
      setInternalChecked(e.target.checked);
    }
    
    // Call external onChange handler if provided
    if (onChange) {
      onChange(e.target.checked);
    }
  };
  
  return (
    <label 
      className={`relative inline-flex items-center cursor-pointer ${disabled ? 'opacity-50 cursor-not-allowed' : ''} ${className}`}
    >
      <input
        type="checkbox"
        className="sr-only"
        checked={isChecked}
        onChange={handleChange}
        disabled={disabled}
        aria-label={ariaLabel}
      />
      <div
        className={`
          ${sizeClasses[size].wrapper}
          ${isChecked ? 'bg-blue-600' : 'bg-gray-200'}
          rounded-full transition-colors duration-200 ease-in-out
        `}
      >
        <div
          className={`
            ${sizeClasses[size].thumb}
            ${isChecked ? sizeClasses[size].thumbTranslate : 'translate-x-0'}
            bg-white rounded-full shadow-md transform transition-transform duration-200 ease-in-out
          `}
        />
      </div>
    </label>
  );
};

export default Switch;
