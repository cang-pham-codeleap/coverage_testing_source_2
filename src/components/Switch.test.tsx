import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Switch from "./Switch";

describe("Switch Component", () => {
  // Test rendering with default props
  it("renders correctly with default props", () => {
    const { container } = render(<Switch />);
    expect(container).toMatchSnapshot();
  });

  // Test controlled state changes
  it("handles controlled state changes", () => {
    const handleChange = jest.fn();
    const { rerender } = render(
      <Switch checked={false} onChange={handleChange} />
    );

    // Find the input and click it
    const input = screen.getByRole("checkbox") as HTMLInputElement;
    fireEvent.click(input);

    // Check if onChange was called with true
    expect(handleChange).toHaveBeenCalledWith(true);

    // Rerender with checked=true to simulate controlled component update
    rerender(<Switch checked={true} onChange={handleChange} />);
    expect(input.checked).toBe(true);
  });

  // Test uncontrolled state behavior
  it("handles uncontrolled state changes", () => {
    render(<Switch defaultChecked={false} />);

    const input = screen.getByRole("checkbox") as HTMLInputElement;
    expect(input.checked).toBe(false);

    // Click the switch
    fireEvent.click(input);
    expect(input.checked).toBe(true);

    // Click again to toggle off
    fireEvent.click(input);
    expect(input.checked).toBe(false);
  });

  // Test disabled state
  it("disables the switch when disabled prop is true", () => {
    const handleChange = jest.fn();
    render(<Switch disabled onChange={handleChange} />);

    const input = screen.getByRole("checkbox") as HTMLInputElement;
    expect(input.disabled).toBe(true);

    // Try to click the disabled switch
    fireEvent.click(input);
    expect(handleChange).toMatchSnapshot();
  });

  // Test different sizes
  it("renders with different sizes", () => {
    const { rerender, container } = render(<Switch size="sm" />);
    expect(container).toMatchSnapshot();

    rerender(<Switch size="md" />);
    expect(container).toMatchSnapshot();

    rerender(<Switch size="lg" />);
    expect(container).toMatchSnapshot();
  });

  // Test with custom class name
  it("applies custom class name", () => {
    render(<Switch className="custom-class" />);
    const label = screen.getByRole("checkbox").closest("label");
    expect(label).toMatchSnapshot;
  });

  // Test accessibility
  it("supports aria-label for accessibility", () => {
    render(<Switch ariaLabel="Toggle feature" />);
    const input = screen.getByRole("checkbox");
    expect(input).toMatchSnapshot();
  });
});
