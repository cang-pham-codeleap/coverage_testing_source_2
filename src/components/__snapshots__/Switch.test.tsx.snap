// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Component disables the switch when disabled prop is true 1`] = `
[MockFunction] {
  "calls": [
    [
      true,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Switch Component renders correctly with default props 1`] = `
<div>
  <label
    class="relative inline-flex items-center cursor-pointer  "
  >
    <input
      class="sr-only"
      type="checkbox"
    />
    <div
      class="
          w-10 h-5
          bg-gray-200
          rounded-full transition-colors duration-200 ease-in-out
        "
    >
      <div
        class="
            w-4 h-4
            translate-x-0
            bg-white rounded-full shadow-md transform transition-transform duration-200 ease-in-out
          "
      />
    </div>
  </label>
</div>
`;

exports[`Switch Component renders with different sizes 1`] = `
<div>
  <label
    class="relative inline-flex items-center cursor-pointer  "
  >
    <input
      class="sr-only"
      type="checkbox"
    />
    <div
      class="
          w-8 h-4
          bg-gray-200
          rounded-full transition-colors duration-200 ease-in-out
        "
    >
      <div
        class="
            w-3 h-3
            translate-x-0
            bg-white rounded-full shadow-md transform transition-transform duration-200 ease-in-out
          "
      />
    </div>
  </label>
</div>
`;

exports[`Switch Component renders with different sizes 2`] = `
<div>
  <label
    class="relative inline-flex items-center cursor-pointer  "
  >
    <input
      class="sr-only"
      type="checkbox"
    />
    <div
      class="
          w-10 h-5
          bg-gray-200
          rounded-full transition-colors duration-200 ease-in-out
        "
    >
      <div
        class="
            w-4 h-4
            translate-x-0
            bg-white rounded-full shadow-md transform transition-transform duration-200 ease-in-out
          "
      />
    </div>
  </label>
</div>
`;

exports[`Switch Component renders with different sizes 3`] = `
<div>
  <label
    class="relative inline-flex items-center cursor-pointer  "
  >
    <input
      class="sr-only"
      type="checkbox"
    />
    <div
      class="
          w-12 h-6
          bg-gray-200
          rounded-full transition-colors duration-200 ease-in-out
        "
    >
      <div
        class="
            w-5 h-5
            translate-x-0
            bg-white rounded-full shadow-md transform transition-transform duration-200 ease-in-out
          "
      />
    </div>
  </label>
</div>
`;

exports[`Switch Component supports aria-label for accessibility 1`] = `
<input
  aria-label="Toggle feature"
  class="sr-only"
  type="checkbox"
/>
`;
